src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,208 @@
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineControl
+    {
+        private IWpfTextView? GetLastActiveIWpfTextView()
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            // If we return null, the calling queue returns and we stop processing.
+            if (ErrorHandler.Failed(_codeWindow.GetLastActiveView(out var textView)))
+                return null;
+
+            return _editorAdaptersFactoryService.GetWpfTextView(textView);
+        }
+
+        /// <summary>
+        /// Starts a new task to compute the data model.
+        /// </summary>
+        private void StartComputeDataModelTask()
+        {
+            // 'true' value is unused. This just signals to the queue that we have work to do.
+            _computeDataModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Makes the LSP document symbol request and creates the data model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolDataModel?> ComputeDataModelAsync(ImmutableSegmentedList<bool> _, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view. This cancellation token controls the entire DocumentOutlineControl
+            // so if we are closed/cancelled on the UI thread, when this jumps back to the UI thread, it will auto-cancel and won't continue
+            // further. We only get to the code below if the control is still in an active state.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var currentSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeModelAsync. It ensures that fetching and processing the document
+            // symbol data model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeModelAsync().ConfigureAwait(false);
+
+            // The model can be null if the LSP document symbol request returns a null response.
+            if (model is not null)
+                StartFilterAndSortDataModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolDataModel?> ComputeModelAsync()
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, _languageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
```

Comment:
i'm curious what hte text buffer is actually needed for.

Reply:
I think this was already discussed in a prior conversation somewhere.

Reply:
Yup, [discussion link](https://github.com/dotnet/roslyn/pull/61784/files/e7e6d6713a13930f068971d7cdd9a541bcb641aa#r909071582) and screenshot below:
![image](https://user-images.githubusercontent.com/32170897/180028147-b5c8c60c-7478-41c1-96ad-668a2a00a4d1.png)